File: day03/src/main/cpp/solution.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <algorithm>
#include <cmath>

#define VLL vector<long long>

using namespace std;

VLL readInputFile() {
    ifstream file("../resources/input");
    VLL lines;
    string s;
    while(getline(file, s)) {
        lines.push_back(stoll(s,0,2));
    }
    return lines;
}

int runPowerDiagnostics(VLL output) {
    int gamma =  0;
    int epsilon = 0;
    int count = 0;
    int highest = (int)(log2(*max_element(output.begin(), output.end())));
    for(int i = 0; i <= highest; i++) {
        count = 0;
        for (auto const &line: output) {
            count += (line >> i) & 1;
        }

        if(count * 2 > output.size()) {
            gamma += (1 << i);
        } else {
            epsilon += (1 << i);
        }
    }
    return gamma * epsilon;
}

int findOutputLine(VLL output, int lsb ) {
 int highest = (int)(log2(*max_element(output.begin(), output.end())));
 int toAdd = 0;
 VLL filtered = output;
 while(filtered.size() > 1) {
     int count = 0;
    for (auto const &line: filtered) {
        count += (line >> highest) & 1;
    }
    if(count * 2 >= filtered.size()) {
        toAdd = lsb;
    } else {
        toAdd = 1 - lsb;
    }
    VLL nextIteration;
    for(auto const &line: filtered) {
        if(((line >> highest) & 1) == toAdd) {
            nextIteration.push_back(line);
        }
    }
    filtered = nextIteration;
    highest--;
 }
 return filtered[0];
}


int runOxygenDiagnostic(VLL output) {
    int oxygen = findOutputLine(output, 1);
    int co2 = findOutputLine(output, 0);
    return oxygen * co2;
}


int main() {
    VLL input = readInputFile();
    cout << "part 1" << endl
         << runPowerDiagnostics(input) << endl
         << "part 2" << endl
         << runOxygenDiagnostic(input) << endl;
    return 0;
}

File: day03/src/main/python/solution.py
import math


def read_input():
    f = open('../resources/input', 'r')
    return list(map(lambda x: int(x, 2), f.readlines()))


def diagnose_power(diagnose):
    gamma = epsilon = 0
    highest = int(math.log2(max(diagnose))) + 1
    for x in range(highest):
        ones = sum(map(lambda a: (a >> x) & 1, diagnose))
        if ones * 2 > len(diagnose):
            gamma = gamma | (1 << x)
        else:
            epsilon = epsilon | (1 << x)
    return gamma * epsilon


def filter_output(diagnose, lsb):
    index = int(math.log2(max(diagnose)))
    current = diagnose.copy()
    while len(current) > 1:
        ones = sum(map(lambda a: (a >> index) & 1, current))
        if ones * 2 >= len(current):
            delete = 1 - lsb
        else:
            delete = lsb
        current = list(filter(lambda a: (a >> index) & 1 == delete, current))
        index = index - 1
    return current[0]


def diagnose_oxygen(diagnose):
    oxygen = filter_output(diagnose, 1)
    co2 = filter_output(diagnose, 0)
    return oxygen * co2


if __name__ == "__main__":
    output = read_input()
    print("part 1")
    print(diagnose_power(output))
    print("part 2")
    print(diagnose_oxygen(output))
